### PR TITLE
Use same lnd's `cert` package to create TLS config to fix TLS cipher suites

### PR DESCRIPTION
**Pull request by guggero**
We need to use safe TLS1.2 cipher suites that are also accepted by the golang `http2` library. Turns out, that is a subset of what we currently use in `lnd`.

### CODE REVIEW

File: aperture.go [hunk 1]
```diff
@@ -189,11 +189,11 @@ func setupLogging(cfg *config) error {
 // getTLSConfig returns a TLS configuration for either a self-signed certificate
 // or one obtained through Let's Encrypt.
 func getTLSConfig(serverName string, autoCert bool) (*tls.Config, error) {
-	var tlsKeyFile, tlsCertFile string
-	switch {
-	// When using autocert, we set a TLSConfig on the server so the key and
-	// cert file we pass in are ignored and don't need to exist.
-	case autoCert:
+	// If the --autocert flag is set, use the autocert library that will
```
Comment (joostjager) [inline aperture.go:192]: It feels a bit strange to mention the exact flag name here.

File: aperture.go [hunk 2]
```diff
@@ -189,11 +189,11 @@ func setupLogging(cfg *config) error {
...
+	// create a new certificate for us from Let's Encrypt as soon as the
```
Comment (joostjager) [inline aperture.go:193]: "create ... from" ? request from?

File: aperture.go [hunk 3]
```diff
@@ -231,6 +186,70 @@ func setupLogging(cfg *config) error {
...
+			}
+			log.Infof("Done generating TLS certificates")
+		}
+
+		certData, _, err := cert.LoadCert(tlsCertFile, tlsKeyFile)
+		if err != nil {
+			return nil, err
+		}
+
+		return cert.TLSConfFromCert(certData), nil
```
Comment (joostjager) [inline aperture.go]: This change is more than only a function extraction, probably deserves a mention in the commit message. Or a separate commit doing this first and the next commit extracting the function.

File: aperture.go [hunk 4]
```diff
@@ -240,11 +240,44 @@ func getTLSConfig(serverName string, autoCert bool) (*tls.Config, error) {
...
 
-	certData, _, err := cert.LoadCert(tlsCertFile, tlsKeyFile)
+	certData, parsedCert, err := cert.LoadCert(tlsCertFile, tlsKeyFile)
 	if err != nil {
 		return nil, err
 	}
 
+	// If the certificate expired or it was outdated, delete it and the TLS
+	// key and generate a new pair.
+	if time.Now().After(parsedCert.NotAfter) {
```
Comment (joostjager) [inline aperture.go:250]: So the workflow is to wait until the certificates has expired and then restart the service? Sounds tricky to do without downtime.

Can't the certificate just never expire?

File: aperture.go (hunk 1 above)
Comment (guggero) [reply aperture.go]: Agree.

File: aperture.go (hunk 2 above)
Comment (guggero) [reply aperture.go]: Fixed the comment to be more precise.

File: aperture.go (hunk 4 above)
Comment (guggero) [reply aperture.go]: This is the exact code that is used in `lnd`. But you are right, that doesn't make a lot of sense for a server that should have as little down time as possible.
No expiration is not possible, most TLS libraries will block a certificate that's valid for too long. But I chose a sane period that even the more restrictive libraries/OSes should allow and set the refresh check to about half that time. So as long as between year 1 and year 2 there is a restart, it should be fine.

**Comment by guggero** [comment]
We ended up disabling TLS to get it to work. There seems to be a general problem with TLS and load balancers.
I decided to leave all the TLS changes in this PR anyway as this still might come in handy later on when we want to run aperture as a Kubernetes Ingress Controller.

File: aperture.go [hunk 5]
```diff
@@ -34,6 +34,18 @@ const (
 	etcdKeyDelimeter = "/"
 )
 
+var (
+	// http2TLSCipherSuites is the list of cipher suites we allow the server
+	// to use. This list removes a CBC cipher from the list used in lnd's
+	// cert package because apparently HTTP/2 and recent go versions are
+	// more strict about what should be allowed.
```
Comment (joostjager) [inline aperture.go:54]: Is there a link to the http2 golang library that defines this?

File: aperture.go [hunk 6]
```diff
@@ -138,6 +152,17 @@ func start() error {
 			}
 			log.Infof("Done generating TLS certificates")
 		}
+
+		// Load the certs now so we can create a complete TLS config.
+		certData, _, err := cert.LoadCert(tlsCertFile, tlsKeyFile)
```
Comment (joostjager) [inline aperture.go]: Cert loading needs to move to the next commit?

File: aperture.go [hunk 7]
```diff
@@ -32,6 +32,19 @@ const (
...
+	// 14 months and is set to a maximum just below what some operating
+	// systems set as a sane maximum certificate duration. See
+	// https://support.apple.com/en-us/HT210176 for more information.
+	selfSignedCertValidity = time.Hour * 24 * 820
+
+	// selfSignedCertExpiryMargin is how much time before the certificate's
+	// expiry date we already refresh it with a new one. We set this to half
+	// the certificate validity length to make the chances bigger for it to
+	// be refreshed on a routine server restart.
+	selfSignedCertExpiryMargin = time.Hour * 24 * 410
```
Comment (joostjager) [inline aperture.go:47]: `selfSignedCertValidity / 2`?

File: aperture.go [hunk 8]
```diff
@@ -247,19 +260,59 @@ func getTLSConfig(serverName string, autoCert bool) (*tls.Config, error) {
...
+	// Load the certs now so we can inspect it and return a complete TLS
+	// config later.
+	certData, parsedCert, err := cert.LoadCert(tlsCertFile, tlsKeyFile)
 	if err != nil {
 		return nil, err
 	}
+
+	// The margin is negative, so adding it to the expiry date should give
+	// us a date in about the middle of it's validity period.
+	expiryWithMargin := parsedCert.NotAfter.Add(
```
Comment (joostjager) [inline aperture.go:292]: `NotAfter.Sub`

File: proxy/proxy.go [hunk 9]
```diff
@@ -70,8 +71,15 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 	defer logRequest()
 
+	// PRI requests are an internal check of HTTP/2 which we just have to
+	// say OK to.
+	if r.Method == "PRI" {
```
Comment (joostjager) [inline proxy/proxy.go:76]: Some more info on this would be nice. What problem do we have if we don't answer?

File: proxy/proxy.go [hunk 10]
```diff
@@ -148,8 +156,8 @@ func (p *Proxy) UpdateServices(services []*Service) error {
 	if err != nil {
 		return err
 	}
-	transport := &http.Transport{
-		ForceAttemptHTTP2: true,
+	transport := &http2.Transport{
+		AllowHTTP: true,
```
Comment (joostjager) [inline proxy/proxy.go:160]: Why this change?

File: proxy/proxy.go [hunk 11]
```diff
@@ -315,6 +323,8 @@ func sendDirectResponse(w http.ResponseWriter, r *http.Request,
 	// Find out if the client is a normal HTTP or a gRPC client. Every gRPC
 	// request should have the Content-Type header field set accordingly
 	// so we can use that.
+	log.Debugf("Sending status code %d for request %s %s %s", statusCode,
+		r.Method, r.RequestURI, r.Proto)
```
Comment (joostjager) [inline proxy/proxy.go:327]: Does this need to stay?

File: aperture.go [hunk 12]
```diff
@@ -113,9 +113,23 @@ func start() error {
 
 	// Create TLS configuration by either creating new self-signed certs or
 	// trying to obtain one through Let's Encrypt.
-	httpsServer.TLSConfig, err = getTLSConfig(cfg.ServerName, cfg.AutoCert)
-	if err != nil {
-		return err
+	var serveFn func() error
+	if cfg.Insecure {
+		serveFn = httpsServer.ListenAndServe
+		httpsServer.Handler = h2c.NewHandler(handler, &http2.Server{})
```
Comment (joostjager) [inline aperture.go:119]: Can also use some documentation about what `h2c` is

File: aperture.go (hunk 6 above)
Comment (guggero) [reply aperture.go]: Not really, we can't create a valid `TLSConfig` without the certificate.

File: aperture.go (hunk 8 above)
Comment (guggero) [reply aperture.go:292]: `Sub` takes a time, not a duration.

File: proxy/proxy.go (hunk 9 above)
Comment (guggero) [reply proxy/proxy.go:76]: Removed the whole commit. We did try out some stuff with this but I think in the end (with TLS turned off, this is not needed anymore). Same with the following 2 commits.

**Comment by guggero** [comment]
We tested this again on the AWS cluster and confirmed it still works.

File: aperture.go (hunk 8 above)
Comment (joostjager) [reply aperture.go:298]: Ah that is right

File: aperture.go [hunk 13]
```diff
@@ -114,9 +114,28 @@ func start() error {
 
 	// Create TLS configuration by either creating new self-signed certs or
 	// trying to obtain one through Let's Encrypt.
-	httpsServer.TLSConfig, err = getTLSConfig(cfg.ServerName, cfg.AutoCert)
-	if err != nil {
-		return err
+	var serveFn func() error
+	if cfg.Insecure {
+		// Normally, HTTP/2 only works with TLS. But there is a special
+		// version called HTTP/2 Cleartext (h2c) that some clients
```
Comment (joostjager) [inline aperture.go:120]: :+1:
